### ISSUE TITLE
Investigate `float` image `nan` support

### ISSUE THREAD

**Issue by Kirill888** [opendatacube/odc-geo#162]
An issue in `datacube` repo highlighted possible problems around `float` image support in `odc-geo` and `datacube`: https://github.com/opendatacube/datacube-core/issues/1595

When creating `float{32,64}` outputs without explicitly configured `nodata`, `odc-geo` uses floating point `nan` value as a `fill_value/nodata`, but `nodata=nan` attribute is not being set on the data-array itself.

Looks like omitting `nodata=nan` when interacting with GDAL causes issues, at least when generating overviews, however reproject might also be impacted. We don't really have a good test suite for that area of the code currently. One probably needs to set `src_nodata=nan, dst_nodata=nan` for  `float{32,64}` arrays even when `nodata` attributes is not set on the array itself.

It could be that GDAL treats `nodata=None` as an *"assume only valid pixel values are present, do not expect to see `nan`s, so go ahead and use optimized code-path"* directive. So, there could be significant performance gains if you know that data contains only valid pixels and you can communicate that to GDAL. We will need a mechanism to distinguish between *"nodata not supplied, use default behaviour for a given pixel type"* and *"user tells us there are NO missing values in the data at all"*. This can be achieved by swapping `nodata=None` for `nodata="auto"` in those functions that allow such overrides and treat `None` value as a *"all pixels are valid in this data"*, ignore `nodata` attribute even if set on the data.

**Comment by SpacemanPaul** [comment]
Also worth noting (from the core issue linked above) that COGs that do not have `nodata=nan` embedded in their metadata are mishandled by ESRI products (ArcGIS, etc).

It's a separate issue I guess, but ideally whatever we end up doing about GDAL should fix the other too.

**Comment by Kirill888** [comment]
Can confirm that reprojecting `float32` data with `nan` pixels present, does indeed require supplying `nodata=nan` to GDAL.

![image](https://github.com/opendatacube/odc-geo/assets/1428024/083570f3-0258-4fcd-b60b-c6048ce7d64f)

```python
import matplotlib.pyplot as plt
import numpy as np
from odc.geo.geobox import GeoBox
from odc.geo.xr import xr_zeros

gbox = GeoBox.from_bbox(
    (-2_000_000, -5_000_000, 2_250_000, -1_000_000),
    "epsg:3577",
    shape=101,
)
gbox2 = gbox.to_crs(3857)

da = xr_zeros(gbox, dtype="float32") + 1
da.data[:, 0:50] = np.nan
da.data[:, 25] = 1
da.data[:, 75] = np.nan
assert da.odc.nodata is None
assert da.odc.geobox == gbox

da1 = da.odc.reproject(gbox2, resampling="average")
da2  = da.odc.reproject(gbox2, resampling="average", src_nodata=float("nan"))


kw = dict(add_labels=False, add_colorbar=False, vmin=0, vmax=1, cmap='Greys')
fig, axs = plt.subplots(1, 3, figsize=(12, 4))

da.plot.imshow(ax=axs[0], **kw)

da1.plot.imshow(ax=axs[1], **kw)
da2.plot.imshow(ax=axs[2], **kw)
for t, ax in zip(
    [
        f"{da.odc.crs.epsg}, Source",
        f"{da1.odc.crs.epsg}, nodata=None",
        f"{da2.odc.crs.epsg}, nodata=NaN",
    ],
    axs,
):
    ax.set_title(t)
```

**Comment by Kirill888** [comment]
Fixed by #163
